### PR TITLE
feat: Support bulk banning in guilds

### PR DESCRIPTION
**Pull request by NeloBlivion**
## Summary

`Guild.ban` now accepts up to 200 members
Currently, this returns two lists:
```py
successes, failures = await guild.ban(*members, reason="Hacked")
```
Alternatively, we could have a new `GuildBulkBan` object with attributes `banned` and `failed`

## Information

- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.
- [x] I have updated the changelog to include these changes.

### CODE REVIEW

File: discord/guild.py [hunk 1]
```diff
@@ -3073,25 +3073,24 @@ async def kick(self, user: Snowflake, *, reason: str | None = None) -> None:
 
     async def ban(
         self,
-        user: Snowflake,
-        *,
+        *users: Snowflake,
         delete_message_seconds: int | None = None,
         delete_message_days: Literal[0, 1, 2, 3, 4, 5, 6, 7] | None = None,
         reason: str | None = None,
     ) -> None:
```
Comment (plun1331) [inline discord/guild.py:3080]: Update typehint

File: CHANGELOG.md [hunk 2]
```diff
@@ -20,6 +20,8 @@ These changes are available on the `master` branch, but have not yet been releas
   ([#2417](https://github.com/Pycord-Development/pycord/pull/2417))
 - Added `Guild.search_members`.
   ([#2418](https://github.com/Pycord-Development/pycord/pull/2418))
+- Added bulk banning up to 200 users through `Guild.ban`
```
Comment (plun1331) [inline CHANGELOG.md:23]:
```suggestion
- Added bulk banning up to 200 users through `Guild.ban`.
```

**Review by plun1331 — CHANGES_REQUESTED** [review]
Possibly make a new method instead (guild.bulk_ban) since they're different endpoints, and similar stuff is split as well (e.g. fetch_member vs fetch_members)

File: discord/guild.py [hunk 3]
```diff
@@ -3073,25 +3073,24 @@ async def kick(self, user: Snowflake, *, reason: str | None = None) -> None:
...
-        """|coro|
+        r"""|coro|
 
-        Bans a user from the guild.
+        Bans users from the guild.
```
Comment (JustaSqu1d) [inline discord/guild.py:3083]:
```suggestion
        Bans user(s) from the guild.
```

File: discord/guild.py [hunk 4]
```diff
@@ -3073,25 +3073,24 @@ async def kick(self, user: Snowflake, *, reason: str | None = None) -> None:
...
 
-        The user must meet the :class:`abc.Snowflake` abc.
+        The users must meet the :class:`abc.Snowflake` abc.
```
Comment (JustaSqu1d) [inline discord/guild.py:3085]:
```suggestion
        The user(s) must meet the :class:`abc.Snowflake` abc.
```

File: discord/guild.py [hunk 5]
```diff
@@ -3121,9 +3138,27 @@ async def ban(
...
+        if len(users) == 1:
+            await self._state.http.ban(
+                users[0].id,
+                self.id,
+                delete_message_seconds,
+                delete_message_days,
+                reason=reason,
+            )
+        elif len(users) > 200:
+            raise ValueError("You may only bulk ban up to 200 members.")
```
Comment (JustaSqu1d) [inline discord/guild.py:3150]: Make the error message more formal; e.g. the number of members that are banned must be less than or equal to 200

File: discord/guild.py [hunk 6]
```diff
@@ -3100,14 +3099,32 @@ async def ban(
...
+        Example Usage: ::
+
+            # Ban a single member
+            await guild.ban(member, reason="Spam")
+
+            # Ban multiple members
+            successes, failures = await guild.ban(member1, member2, member3, ..., reason="Raid")
+
+            # Ban a list of members
+            successes, failures = await guild.ban(*members, reason="Hacked")
```
Comment (JustaSqu1d) [inline discord/guild.py:3127]: This is formatted weirdly on docs

image

**Review by JustaSqu1d — CHANGES_REQUESTED** [review]

File: discord/guild.py (hunk 1 above)
Comment (JustaSqu1d) [reply discord/guild.py:3080]: not sure which type-hint this refers to, but also the method's return value.

Comment (plun1331) [reply discord/guild.py:3080]: > not sure which type-hint this refers to, but also the method's return value.

That's what I was referring to

File: discord/guild.py (hunk 6 above)
Comment (plun1331) [reply discord/guild.py:3127]: Probably needs to be put above params n stuff

**Comment by NeloBlivion** [comment]
> Possibly make a new method instead (guild.bulk_ban) since they're different endpoints, and similar stuff is split as well (e.g. fetch_member vs fetch_members)

I did consider this initially, but after mulling it over I figured it made more sense to compare to `Member.add_roles/remove_roles`, which similarly accepts 1 or multiple Snowflakes and incorporates different routes depending on the situation

**Comment by plun1331** [comment]
> > Possibly make a new method instead (guild.bulk_ban) since they're different endpoints, and similar stuff is split as well (e.g. fetch_member vs fetch_members)
> 
> I did consider this initially, but after mulling it over I figured it made more sense to compare to `Member.add_roles/remove_roles`, which similarly accepts 1 or multiple Snowflakes and incorporates different routes depending on the situation

My main issue is we end up having multiple different return types, which isn't a massive problem but might be confusing as some people might pass a list and not expect it to give `None` back (if the list had a length of 1), similar to some confusion that was discussed a while ago about the return types of `ApplicationContext.respond`

**Review by Lulalaby — APPROVED** [review]

**Comment by Dorukyum** [comment]
I agree with plun, a new method should suffice.

**Comment by Lulalaby** [comment]
Upon implementing it myself I agree with doru, a seperate method is needed

**Review by Lulalaby — DISMISSED** [review]
Revert changes and introduce bulk banning as new method bulk_ban.
Also we should finally remove delete message days

**Comment by NeloBlivion** [comment]
With this, 2.6 is officially breaking so please include the relevant warnings in any announcements

**Review by plun1331 — APPROVED** [review]

**Review by Lulalaby — APPROVED** [review]

**Comment by Lulalaby** [comment]
Before we merge it, @Dorukyum, do you agree on removing the obsolete parameter completely in this PR?

**Comment by Dorukyum** [comment]
> Before we merge it, @Dorukyum, do you agree on removing the obsolete parameter completely in this PR?

It's fine. A seperate pr for each change would obviously be better, but the changes here are relevant enough I guess.

**Review by Lulalaby — APPROVED** [review]

**Pull request merged**